// @rocket.chat/onboarding-ui/dist/cjs/forms/RegisterServerForm/RegisterServerForm.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._RegisterServerFormOffline = exports._RegisterServerForm = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var RegisterServerForm_1 = __importDefault(require("./RegisterServerForm"));
exports.default = {
    title: 'forms/RegisterServerForm',
    component: RegisterServerForm_1.default,
    parameters: {
        layout: 'centered',
        actions: { argTypesRegex: '^on.*' },
    },
    args: {
        currentStep: 1,
        stepCount: 1,
    },
};
var _RegisterServerForm = function (args) { return ((0, jsx_runtime_1.jsx)(RegisterServerForm_1.default, __assign({}, args))); };
exports._RegisterServerForm = _RegisterServerForm;
exports._RegisterServerForm.storyName = 'RegisterServerForm';
var _RegisterServerFormOffline = function (args) { return ((0, jsx_runtime_1.jsx)(RegisterServerForm_1.default, __assign({}, args, { offline: true }))); };
exports._RegisterServerFormOffline = _RegisterServerFormOffline;
exports._RegisterServerFormOffline.storyName = 'RegisterServerFormOffline';
