// @rocket.chat/onboarding-ui/dist/esm/common/FormPageLayout.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Tile } from '@rocket.chat/fuselage';
import FormPageLayout from './FormPageLayout';
export default {
    title: 'common/FormPageLayout',
    component: FormPageLayout,
    parameters: {
        layout: 'fullscreen',
        actions: { argTypesRegex: '^on.*' },
    },
    args: {
        title: _jsx(Tile, { children: "Title" }),
        subtitle: _jsx(Tile, { children: "Subtitle" }),
        description: _jsx(Tile, { children: "Description" }),
        children: _jsx(Tile, { children: "Form" }),
    },
};
export var _FormPageLayout = function (args) { return (_jsx(FormPageLayout, __assign({}, args))); };
_FormPageLayout.storyName = 'FormPageLayout';
