// @rocket.chat/layout/dist/cjs/BackgroundLayer/BackgroundLayer.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightMode = exports.DarkMode = exports.SystemDarkMode = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var fuselage_1 = require("@rocket.chat/fuselage");
var BackgroundLayer_1 = __importDefault(require("./BackgroundLayer"));
exports.default = {
    title: 'common/BackgroundLayer',
    component: BackgroundLayer_1.default,
    parameters: {
        layout: 'fullscreen',
    },
};
var SystemDarkMode = function () { return ((0, jsx_runtime_1.jsx)(BackgroundLayer_1.default, { children: (0, jsx_runtime_1.jsx)(fuselage_1.Tile, { children: "An example tile" }) })); };
exports.SystemDarkMode = SystemDarkMode;
var DarkMode = function (props) { return ((0, jsx_runtime_1.jsx)(BackgroundLayer_1.default, __assign({}, props, { children: (0, jsx_runtime_1.jsx)(fuselage_1.Tile, { children: "An example tile" }) }))); };
exports.DarkMode = DarkMode;
exports.DarkMode.args = { forcedDarkMode: true };
var LightMode = function (props) { return ((0, jsx_runtime_1.jsx)(BackgroundLayer_1.default, __assign({}, props, { children: (0, jsx_runtime_1.jsx)(fuselage_1.Tile, { children: "An example tile" }) }))); };
exports.LightMode = LightMode;
exports.LightMode.args = { forcedDarkMode: false };
