// @rocket.chat/onboarding-ui/dist/esm/pages/LoaderPage/LoaderPage.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import LoaderPage from './LoaderPage';
var subtitles = [
    'Bringing rocket to launch position1  🚀',
    'Loading rocket propellant  🚀',
    'Performing final go/no go poll with flight crew  🚀',
    'All systems nominal, switching to internal computer  🚀',
    'Beginning countdown, 5...4...3...2...1  🚀',
    'Transitioning to liftoff  🚀',
];
export default {
    title: 'pages/LoaderPage',
    component: LoaderPage,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        title: 'Launching Kapai...',
        subtitles: subtitles,
        isReady: false,
        loadingSeconds: 15,
    },
};
export var _LoaderPage = function (args) { return _jsx(LoaderPage, __assign({}, args)); };
_LoaderPage.storyName = 'LoaderPage';
