// @rocket.chat/license/dist/pendingLicense.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearPendingLicense = exports.isPendingLicense = exports.hasPendingLicense = exports.applyPendingLicense = exports.setPendingLicense = void 0;
const logger_1 = require("./logger");
function setPendingLicense(encryptedLicense) {
    this.pendingLicense = encryptedLicense;
    if (this.pendingLicense) {
        logger_1.logger.info('Storing license as pending validation.');
    }
}
exports.setPendingLicense = setPendingLicense;
async function applyPendingLicense() {
    if (this.pendingLicense) {
        logger_1.logger.info('Applying pending license.');
        return this.setLicense(this.pendingLicense);
    }
}
exports.applyPendingLicense = applyPendingLicense;
function hasPendingLicense() {
    return Boolean(this.pendingLicense);
}
exports.hasPendingLicense = hasPendingLicense;
function isPendingLicense(encryptedLicense) {
    return !!this.pendingLicense && this.pendingLicense === encryptedLicense;
}
exports.isPendingLicense = isPendingLicense;
function clearPendingLicense() {
    if (this.pendingLicense) {
        logger_1.logger.info('Removing pending license.');
    }
    this.pendingLicense = '';
}
exports.clearPendingLicense = clearPendingLicense;
