// @rocket.chat/onboarding-ui/dist/cjs/pages/OauthAuthorizationPage/OauthAuthorizationPage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._OauthAuthorizationPage = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var OauthAuthorizationPage_1 = __importDefault(require("./OauthAuthorizationPage"));
exports.default = {
    title: 'pages/OauthAuthorizationPage',
    component: OauthAuthorizationPage_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        clientName: 'Rocket.Chat',
        error: {
            message: '',
            onGoBack: function () { return console.log('Back'); },
        },
    },
};
var _OauthAuthorizationPage = function (args) { return ((0, jsx_runtime_1.jsx)(OauthAuthorizationPage_1.default, __assign({}, args))); };
exports._OauthAuthorizationPage = _OauthAuthorizationPage;
exports._OauthAuthorizationPage.storyName = 'OauthAuthorizationPage';
