// @rocket.chat/onboarding-ui/dist/cjs/pages/RegisterServerPage/RegisterServerPage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._RegisterServerPageOffline = exports._RegisterServerPage = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var RegisterServerPage_1 = __importDefault(require("./RegisterServerPage"));
exports.default = {
    title: 'pages/RegisterServerPage',
    component: RegisterServerPage_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        currentStep: 1,
        stepCount: 1,
        termsHref: '#',
        policyHref: '#',
    },
};
var _RegisterServerPage = function (args) { return ((0, jsx_runtime_1.jsx)(RegisterServerPage_1.default, __assign({}, args))); };
exports._RegisterServerPage = _RegisterServerPage;
exports._RegisterServerPage.storyName = 'Online';
var _RegisterServerPageOffline = function (args) { return ((0, jsx_runtime_1.jsx)(RegisterServerPage_1.default, __assign({}, args, { offline: true }))); };
exports._RegisterServerPageOffline = _RegisterServerPageOffline;
exports._RegisterServerPageOffline.storyName = 'Offline';
