// @rocket.chat/onboarding-ui/dist/cjs/pages/LoaderPage/LoaderPage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._LoaderPage = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var LoaderPage_1 = __importDefault(require("./LoaderPage"));
var subtitles = [
    'Bringing rocket to launch position1  🚀',
    'Loading rocket propellant  🚀',
    'Performing final go/no go poll with flight crew  🚀',
    'All systems nominal, switching to internal computer  🚀',
    'Beginning countdown, 5...4...3...2...1  🚀',
    'Transitioning to liftoff  🚀',
];
exports.default = {
    title: 'pages/LoaderPage',
    component: LoaderPage_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        title: 'Launching Kapai...',
        subtitles: subtitles,
        isReady: false,
        loadingSeconds: 15,
    },
};
var _LoaderPage = function (args) { return (0, jsx_runtime_1.jsx)(LoaderPage_1.default, __assign({}, args)); };
exports._LoaderPage = _LoaderPage;
exports._LoaderPage.storyName = 'LoaderPage';
