// @rocket.chat/onboarding-ui/dist/esm/forms/AwaitConfirmationForm/AwaitConfirmationForm.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import AwaitConfirmationForm from './AwaitConfirmationForm';
export default {
    title: 'forms/AwaitConfirmationForm',
    component: AwaitConfirmationForm,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'centered',
    },
    argTypes: {
        onResendEmailRequest: { action: 'resetEmailRequest' },
        onChangeEmailRequest: { action: 'changeEmailRequest' },
    },
    args: {
        currentStep: 4,
        stepCount: 4,
        securityCode: 'Funny Tortoise In The Hat',
        emailAddress: 'emailname@email.com',
    },
};
export var _AwaitConfirmationForm = function (args) { return (_jsx(AwaitConfirmationForm, __assign({}, args))); };
_AwaitConfirmationForm.storyName = 'AwaitConfirmationForm';
