// @rocket.chat/onboarding-ui/dist/cjs/forms/CreateCloudWorkspaceForm/CreateCloudWorkspaceForm.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._WorkspaceUrlInput = exports._CreateCloudWorkspaceForm = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var CreateCloudWorkspaceForm_1 = __importDefault(require("./CreateCloudWorkspaceForm"));
var WorkspaceUrlInput_1 = __importDefault(require("./WorkspaceUrlInput"));
var isValidLength = function (domainName) {
    if (domainName.length < 3) {
        return 'Workspace URL should have at least 3 characters';
    }
    return true;
};
var isValidFirstSubstring = function (domainName) {
    var firstSubstring = domainName.slice(0, 3);
    if (firstSubstring === 'rc-') {
        return 'Workspace URL address unavailable';
    }
    return true;
};
var isValidCharacter = function (domainName) {
    var regex = RegExp('^([a-zA-Z0-9]([a-zA-Z0-9\\-]{0,61}[a-zA-Z0-9]))$');
    if (!regex.test(domainName)) {
        return 'Invalid character for Workspace URL';
    }
    return true;
};
var isValidDomainName = function (domainName) { return __awaiter(void 0, void 0, void 0, function () {
    var validationList, _i, validationList_1, validate, validateMessage;
    return __generator(this, function (_a) {
        validationList = [
            isValidLength,
            isValidFirstSubstring,
            isValidCharacter,
        ];
        for (_i = 0, validationList_1 = validationList; _i < validationList_1.length; _i++) {
            validate = validationList_1[_i];
            validateMessage = validate(domainName);
            if (typeof validateMessage === 'string') {
                return [2 /*return*/, validateMessage];
            }
        }
        return [2 /*return*/, true];
    });
}); };
exports.default = {
    title: 'forms/CreateCloudWorkspaceForm',
    component: CreateCloudWorkspaceForm_1.default,
    parameters: {
        layout: 'centered',
        actions: { argTypesRegex: '^on.*' },
    },
    args: {
        serverRegionOptions: [
            ['us', 'US'],
            ['br', 'BR'],
        ],
        languageOptions: [
            ['en', 'English'],
            ['pt', 'Português'],
        ],
        domain: 'rocket.chat',
        validateUrl: isValidDomainName,
        validateEmail: function (email) { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
            return [2 /*return*/, email === 'rocket@rocket.chat' ? 'invalid email' : true];
        }); }); },
    },
};
var _CreateCloudWorkspaceForm = function (args) { return ((0, jsx_runtime_1.jsx)(CreateCloudWorkspaceForm_1.default, __assign({}, args))); };
exports._CreateCloudWorkspaceForm = _CreateCloudWorkspaceForm;
var _WorkspaceUrlInput = function () { return ((0, jsx_runtime_1.jsx)(WorkspaceUrlInput_1.default, { domain: 'rocket.chat' })); };
exports._WorkspaceUrlInput = _WorkspaceUrlInput;
exports._CreateCloudWorkspaceForm.storyName = 'CreateCloudWorkspaceForm';
