// @rocket.chat/onboarding-ui/dist/esm/forms/NewAccountForm/NewAccountForm.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import NewAccountForm from './NewAccountForm';
export default {
    title: 'forms/NewAccountForm',
    component: NewAccountForm,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'centered',
    },
    argTypes: {
        validateUsername: { action: 'validateUsername' },
        validateEmail: { action: 'validateEmail' },
        validatePassword: { action: 'validatePassword' },
        validateConfirmPassword: { action: 'validateConfirmPassword' },
    },
    args: {},
};
export var _NewAccountForm = function (args) { return (_jsx(NewAccountForm, __assign({}, args))); };
_NewAccountForm.storyName = 'NewAccountForm';
