// @rocket.chat/rest-typings/dist/v1/channels/ChannelsGetIntegrationsProps.js
import Ajv from 'ajv';
const ajv = new Ajv({
    coerceTypes: true,
});
const channelsGetIntegrationsPropsSchema = {
    oneOf: [
        {
            type: 'object',
            properties: {
                roomId: {
                    type: 'string',
                },
                includeAllPublicChannels: {
                    type: 'string',
                    nullable: true,
                },
            },
            required: ['roomId'],
            additionalProperties: false,
        },
        {
            type: 'object',
            properties: {
                roomName: {
                    type: 'string',
                },
                includeAllPublicChannels: {
                    type: 'string',
                    nullable: true,
                },
            },
            required: ['roomName'],
            additionalProperties: false,
        },
    ],
};
export const isChannelsGetIntegrationsProps = ajv.compile(channelsGetIntegrationsPropsSchema);
