// @rocket.chat/onboarding-ui/dist/cjs/pages/CheckYourEmailPage/CheckYourEmailPage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._CheckYourEmailPage = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var addon_actions_1 = require("@storybook/addon-actions");
var CheckYourEmailPage_1 = __importDefault(require("./CheckYourEmailPage"));
exports.default = {
    title: 'pages/CheckYourEmailPage',
    component: CheckYourEmailPage_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        title: '',
        children: '',
        onResendEmailRequest: (0, addon_actions_1.action)('onResendEmailRequest'),
        onChangeEmailRequest: (0, addon_actions_1.action)('onChangeEmailRequest'),
    },
};
var _CheckYourEmailPage = function (args) { return ((0, jsx_runtime_1.jsx)(CheckYourEmailPage_1.default, __assign({}, args))); };
exports._CheckYourEmailPage = _CheckYourEmailPage;
exports._CheckYourEmailPage.storyName = 'CheckYourEmailPage';
