// @rocket.chat/onboarding-ui/dist/cjs/forms/RegisterOfflineForm/RegisterOfflineForm.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._RegisterOfflineForm = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var RegisterOfflineForm_1 = __importDefault(require("./RegisterOfflineForm"));
exports.default = {
    title: 'forms/RegisterOfflineForm',
    component: RegisterOfflineForm_1.default,
    parameters: {
        layout: 'centered',
        actions: { argTypesRegex: '^on.*' },
    },
    args: {
        clientKey: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
    },
};
var _RegisterOfflineForm = function (args) { return ((0, jsx_runtime_1.jsx)(RegisterOfflineForm_1.default, __assign({}, args))); };
exports._RegisterOfflineForm = _RegisterOfflineForm;
