// @rocket.chat/onboarding-ui/dist/cjs/pages/AdminInfoPage/AdminInfoPage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._CloudAdminInfoPage = exports._AdminInfoPage = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var AdminInfoPage_1 = __importDefault(require("./AdminInfoPage"));
exports.default = {
    title: 'pages/AdminInfoPage',
    component: AdminInfoPage_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    argTypes: {
        validateUsername: { action: 'validateUsername' },
        validateEmail: { action: 'validateEmail' },
        validatePassword: { action: 'validatePassword' },
    },
    args: {
        currentStep: 1,
        stepCount: 1,
        passwordRulesHint: 'Password rules description goes here',
        control: { type: 'inline-radio' },
    },
};
var _AdminInfoPage = function (args) { return ((0, jsx_runtime_1.jsx)(AdminInfoPage_1.default, __assign({}, args))); };
exports._AdminInfoPage = _AdminInfoPage;
var _CloudAdminInfoPage = function (args) { return ((0, jsx_runtime_1.jsx)(AdminInfoPage_1.default, __assign({}, args))); };
exports._CloudAdminInfoPage = _CloudAdminInfoPage;
exports._AdminInfoPage.storyName = 'AdminInfoPage';
exports._CloudAdminInfoPage.storyName = 'CloudAdminInfoPage';
exports._CloudAdminInfoPage.args = {
    title: 'Your Workspace is Ready!',
    description: 'Time to create your first admin user',
    keepPosted: true,
};
