// @rocket.chat/onboarding-ui/dist/cjs/common/FormPageLayout.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._FormPageLayout = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var fuselage_1 = require("@rocket.chat/fuselage");
var FormPageLayout_1 = __importDefault(require("./FormPageLayout"));
exports.default = {
    title: 'common/FormPageLayout',
    component: FormPageLayout_1.default,
    parameters: {
        layout: 'fullscreen',
        actions: { argTypesRegex: '^on.*' },
    },
    args: {
        title: (0, jsx_runtime_1.jsx)(fuselage_1.Tile, { children: "Title" }),
        subtitle: (0, jsx_runtime_1.jsx)(fuselage_1.Tile, { children: "Subtitle" }),
        description: (0, jsx_runtime_1.jsx)(fuselage_1.Tile, { children: "Description" }),
        children: (0, jsx_runtime_1.jsx)(fuselage_1.Tile, { children: "Form" }),
    },
};
var _FormPageLayout = function (args) { return ((0, jsx_runtime_1.jsx)(FormPageLayout_1.default, __assign({}, args))); };
exports._FormPageLayout = _FormPageLayout;
exports._FormPageLayout.storyName = 'FormPageLayout';
