// @rocket.chat/onboarding-ui/dist/cjs/common/InformationTooltipTrigger.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._InformationTooltipTrigger = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var InformationTooltipTrigger_1 = __importDefault(require("./InformationTooltipTrigger"));
exports.default = {
    title: 'common/InformationTooltipTrigger',
    component: InformationTooltipTrigger_1.default,
    argTypes: {
        text: {
            control: { type: 'text' },
            defaultValue: 'Some tooltip text',
        },
    },
};
var _InformationTooltipTrigger = function (args) { return ((0, jsx_runtime_1.jsx)(InformationTooltipTrigger_1.default, { text: args.text })); };
exports._InformationTooltipTrigger = _InformationTooltipTrigger;
exports._InformationTooltipTrigger.storyName = 'InformationTooltipTrigger';
exports._InformationTooltipTrigger.parameters = {
    layout: 'centered',
};
