// @rocket.chat/rest-typings/dist/v1/users/UsersUpdateOwnBasicInfoParamsPOST.js
import Ajv from 'ajv';
const ajv = new Ajv({
    coerceTypes: true,
});
const UsersUpdateOwnBasicInfoParamsPostSchema = {
    type: 'object',
    properties: {
        data: {
            type: 'object',
            properties: {
                email: {
                    type: 'string',
                    nullable: true,
                },
                name: {
                    type: 'string',
                    nullable: true,
                },
                username: {
                    type: 'string',
                    nullable: true,
                },
                nickname: {
                    type: 'string',
                    nullable: true,
                },
                bio: {
                    type: 'string',
                    nullable: true,
                },
                statusType: {
                    type: 'string',
                    nullable: true,
                },
                statusText: {
                    type: 'string',
                    nullable: true,
                },
                currentPassword: {
                    type: 'string',
                    nullable: true,
                },
                newPassword: {
                    type: 'string',
                    nullable: true,
                },
            },
            required: [],
            additionalProperties: false,
        },
        customFields: {
            type: 'object',
            nullable: true,
        },
    },
    required: ['data'],
    additionalProperties: false,
};
export const isUsersUpdateOwnBasicInfoParamsPOST = ajv.compile(UsersUpdateOwnBasicInfoParamsPostSchema);
