// @rocket.chat/onboarding-ui/dist/esm/forms/AdminInfoForm/AdminInfoForm.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import AdminInfoForm from './AdminInfoForm';
export default {
    title: 'forms/AdminInfoForm',
    component: AdminInfoForm,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'centered',
    },
    argTypes: {
        validateUsername: { action: 'validateUsername' },
        validateEmail: { action: 'validateEmail' },
        validatePassword: { action: 'validatePassword' },
    },
    args: {
        currentStep: 1,
        stepCount: 1,
        passwordRulesHint: 'Password rules description goes here',
    },
};
export var _AdminInfoForm = function (args) { return (_jsx(AdminInfoForm, __assign({}, args))); };
_AdminInfoForm.storyName = 'AdminInfoForm';
