// @rocket.chat/pdf-worker/dist/templates/ChatTranscript/markup/elements/CodeSpan.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.codeStyles = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const renderer_1 = require("@react-pdf/renderer");
const colors_json_1 = __importDefault(require("@rocket.chat/fuselage-tokens/colors.json"));
exports.codeStyles = renderer_1.StyleSheet.create({
    wrapper: {
        backgroundColor: colors_json_1.default.n100,
        borderColor: colors_json_1.default.n250,
        borderWidth: 1,
        borderRadius: 4,
        paddingHorizontal: 3,
        paddingVertical: 1,
    },
    space: {
        backgroundColor: colors_json_1.default.n100,
    },
    code: {
        backgroundColor: colors_json_1.default.n100,
        color: colors_json_1.default.n800,
        fontSize: 13,
        fontWeight: 700,
        fontFamily: 'FiraCode',
    },
});
const CodeSpan = ({ code }) => ((0, jsx_runtime_1.jsxs)(renderer_1.Text, { style: exports.codeStyles.wrapper, children: [(0, jsx_runtime_1.jsx)(renderer_1.Text, { style: exports.codeStyles.space, children: " " }), (0, jsx_runtime_1.jsx)(renderer_1.Text, { style: exports.codeStyles.code, children: code }), (0, jsx_runtime_1.jsx)(renderer_1.Text, { style: exports.codeStyles.space, children: " " })] }));
exports.default = CodeSpan;
