// @rocket.chat/onboarding-ui/dist/cjs/pages/AwaitingConfirmationPage/AwaitingConfirmationPage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._AwaitingConfirmationPage = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var AwaitingConfirmationPage_1 = __importDefault(require("./AwaitingConfirmationPage"));
exports.default = {
    title: 'pages/AwaitingConfirmationPage',
    component: AwaitingConfirmationPage_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    argTypes: {
        onResendEmailRequest: { action: 'resetEmailRequest' },
        onChangeEmailRequest: { action: 'changeEmailRequest' },
    },
    args: {
        currentStep: 4,
        stepCount: 4,
        securityCode: 'Funny Tortoise In The Hat',
        emailAddress: 'emailname@email.com',
    },
};
var _AwaitingConfirmationPage = function (args) { return ((0, jsx_runtime_1.jsx)(AwaitingConfirmationPage_1.default, __assign({}, args))); };
exports._AwaitingConfirmationPage = _AwaitingConfirmationPage;
exports._AwaitingConfirmationPage.storyName = 'AwaitingConfirmationPage';
