// @rocket.chat/onboarding-ui/dist/cjs/pages/InformationPage/InformationPage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._InformationPage = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var InformationPage_1 = __importDefault(require("./InformationPage"));
exports.default = {
    title: 'pages/InformationPage',
    component: InformationPage_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        title: 'Place your title here',
        description: 'Place your description here',
    },
};
var _InformationPage = function (args) { return ((0, jsx_runtime_1.jsx)(InformationPage_1.default, __assign({}, args))); };
exports._InformationPage = _InformationPage;
exports._InformationPage.storyName = 'InformationPage';
