// @rocket.chat/pdf-worker/dist/templates/ChatTranscript/markup/blocks/UnorderedListBlock.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const renderer_1 = require("@react-pdf/renderer");
const InlineElements_1 = __importDefault(require("../elements/InlineElements"));
const styles = renderer_1.StyleSheet.create({
    wrapper: {
        marginTop: 4,
    },
    list: {
        display: 'flex',
        flexDirection: 'row',
    },
    bullet: {
        marginHorizontal: 4,
    },
});
const UnorderedListBlock = ({ items }) => ((0, jsx_runtime_1.jsx)(renderer_1.View, { style: styles.wrapper, children: items.map(({ value }, index) => ((0, jsx_runtime_1.jsxs)(renderer_1.View, { style: styles.list, children: [(0, jsx_runtime_1.jsx)(renderer_1.Text, { style: styles.bullet, children: "\u2022" }), (0, jsx_runtime_1.jsx)(InlineElements_1.default, { children: value })] }, index))) }));
exports.default = UnorderedListBlock;
