// @rocket.chat/onboarding-ui/dist/cjs/pages/ResetPasswordPage/ResetPasswordPage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._ResetPasswordPage = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var ResetPasswordPage_1 = __importDefault(require("./ResetPasswordPage"));
exports.default = {
    title: 'pages/ResetPasswordPage',
    component: ResetPasswordPage_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    argTypes: {
        validateEmail: { action: 'validateEmail' },
    },
    args: {
        control: { type: 'inline-radio' },
    },
};
var _ResetPasswordPage = function (args) { return ((0, jsx_runtime_1.jsx)(ResetPasswordPage_1.default, __assign({}, args))); };
exports._ResetPasswordPage = _ResetPasswordPage;
exports._ResetPasswordPage.storyName = 'ResetPasswordPage';
