// @rocket.chat/onboarding-ui/dist/cjs/pages/CreateFirstMemberPage/CreateFirstMemberPage.spec.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var jsx_runtime_1 = require("react/jsx-runtime");
var react_dom_1 = __importDefault(require("react-dom"));
var CreateFirstMemberPage_1 = __importDefault(require("./CreateFirstMemberPage"));
var onSubmit = jest.fn();
var onBackButtonClick = jest.fn();
var validateUsername = jest.fn();
var validatePassword = jest.fn();
var props = {
    currentStep: 1,
    stepCount: 1,
    organizationName: 'Kapai',
    onSubmit: onSubmit,
    onBackButtonClick: onBackButtonClick,
    validateUsername: validateUsername,
    validatePassword: validatePassword,
};
it('renders without crashing', function () {
    var div = document.createElement('div');
    react_dom_1.default.render((0, jsx_runtime_1.jsx)(CreateFirstMemberPage_1.default, __assign({}, props)), div);
    react_dom_1.default.unmountComponentAtNode(div);
});
