// @rocket.chat/onboarding-ui/dist/esm/pages/CheckYourEmailPage/CheckYourEmailPage.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { action } from '@storybook/addon-actions';
import CheckYourEmailPage from './CheckYourEmailPage';
export default {
    title: 'pages/CheckYourEmailPage',
    component: CheckYourEmailPage,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        title: '',
        children: '',
        onResendEmailRequest: action('onResendEmailRequest'),
        onChangeEmailRequest: action('onChangeEmailRequest'),
    },
};
export var _CheckYourEmailPage = function (args) { return (_jsx(CheckYourEmailPage, __assign({}, args))); };
_CheckYourEmailPage.storyName = 'CheckYourEmailPage';
