// @rocket.chat/core-services/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnterpriseSettings = exports.Import = exports.OmnichannelEEService = exports.Omnichannel = exports.FederationEE = exports.Federation = exports.OmnichannelIntegration = exports.Settings = exports.Translation = exports.Message = exports.OmnichannelTranscript = exports.QueueWorker = exports.Calendar = exports.Upload = exports.VideoConf = exports.DeviceManagement = exports.SAUMonitor = exports.LDAP = exports.Analytics = exports.LivechatVoip = exports.Voip = exports.Media = exports.Room = exports.MessageReads = exports.Team = exports.NPS = exports.UiKitCoreApp = exports.Banner = exports.MeteorService = exports.License = exports.Account = exports.Presence = exports.Apps = exports.Authorization = exports.proxifyWithWait = exports.ServiceClassInternal = exports.ServiceClass = exports.LocalBroker = exports.api = exports.isMeteorError = exports.MeteorError = exports.asyncLocalStorage = void 0;
const proxify_1 = require("./lib/proxify");
Object.defineProperty(exports, "proxifyWithWait", { enumerable: true, get: function () { return proxify_1.proxifyWithWait; } });
var asyncLocalStorage_1 = require("./lib/asyncLocalStorage");
Object.defineProperty(exports, "asyncLocalStorage", { enumerable: true, get: function () { return asyncLocalStorage_1.asyncLocalStorage; } });
var MeteorError_1 = require("./MeteorError");
Object.defineProperty(exports, "MeteorError", { enumerable: true, get: function () { return MeteorError_1.MeteorError; } });
Object.defineProperty(exports, "isMeteorError", { enumerable: true, get: function () { return MeteorError_1.isMeteorError; } });
var api_1 = require("./api");
Object.defineProperty(exports, "api", { enumerable: true, get: function () { return api_1.api; } });
var LocalBroker_1 = require("./LocalBroker");
Object.defineProperty(exports, "LocalBroker", { enumerable: true, get: function () { return LocalBroker_1.LocalBroker; } });
var ServiceClass_1 = require("./types/ServiceClass");
Object.defineProperty(exports, "ServiceClass", { enumerable: true, get: function () { return ServiceClass_1.ServiceClass; } });
Object.defineProperty(exports, "ServiceClassInternal", { enumerable: true, get: function () { return ServiceClass_1.ServiceClassInternal; } });
// TODO think in a way to not have to pass the service name to proxify here as well
exports.Authorization = (0, proxify_1.proxifyWithWait)('authorization');
exports.Apps = (0, proxify_1.proxifyWithWait)('apps-engine');
exports.Presence = (0, proxify_1.proxifyWithWait)('presence');
exports.Account = (0, proxify_1.proxifyWithWait)('accounts');
exports.License = (0, proxify_1.proxifyWithWait)('license');
exports.MeteorService = (0, proxify_1.proxifyWithWait)('meteor');
exports.Banner = (0, proxify_1.proxifyWithWait)('banner');
exports.UiKitCoreApp = (0, proxify_1.proxifyWithWait)('uikit-core-app');
exports.NPS = (0, proxify_1.proxifyWithWait)('nps');
exports.Team = (0, proxify_1.proxifyWithWait)('team');
exports.MessageReads = (0, proxify_1.proxifyWithWait)('message-reads');
exports.Room = (0, proxify_1.proxifyWithWait)('room');
exports.Media = (0, proxify_1.proxifyWithWait)('media');
exports.Voip = (0, proxify_1.proxifyWithWait)('voip');
exports.LivechatVoip = (0, proxify_1.proxifyWithWait)('omnichannel-voip');
exports.Analytics = (0, proxify_1.proxifyWithWait)('analytics');
exports.LDAP = (0, proxify_1.proxifyWithWait)('ldap');
exports.SAUMonitor = (0, proxify_1.proxifyWithWait)('sau-monitor');
exports.DeviceManagement = (0, proxify_1.proxifyWithWait)('device-management');
exports.VideoConf = (0, proxify_1.proxifyWithWait)('video-conference');
exports.Upload = (0, proxify_1.proxifyWithWait)('upload');
exports.Calendar = (0, proxify_1.proxifyWithWait)('calendar');
exports.QueueWorker = (0, proxify_1.proxifyWithWait)('queue-worker');
exports.OmnichannelTranscript = (0, proxify_1.proxifyWithWait)('omnichannel-transcript');
exports.Message = (0, proxify_1.proxifyWithWait)('message');
exports.Translation = (0, proxify_1.proxifyWithWait)('translation');
exports.Settings = (0, proxify_1.proxifyWithWait)('settings');
exports.OmnichannelIntegration = (0, proxify_1.proxifyWithWait)('omnichannel-integration');
exports.Federation = (0, proxify_1.proxifyWithWait)('federation');
exports.FederationEE = (0, proxify_1.proxifyWithWait)('federation-enterprise');
exports.Omnichannel = (0, proxify_1.proxifyWithWait)('omnichannel');
exports.OmnichannelEEService = (0, proxify_1.proxifyWithWait)('omnichannel-ee');
exports.Import = (0, proxify_1.proxifyWithWait)('import');
// Calls without wait. Means that the service is optional and the result may be an error
// of service/method not available
exports.EnterpriseSettings = (0, proxify_1.proxify)('ee-settings');
