// @rocket.chat/onboarding-ui/dist/esm/pages/OauthAuthorizationPage/OauthAuthorizationPage.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import OauthAuthorizationPage from './OauthAuthorizationPage';
export default {
    title: 'pages/OauthAuthorizationPage',
    component: OauthAuthorizationPage,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        clientName: 'Rocket.Chat',
        error: {
            message: '',
            onGoBack: function () { return console.log('Back'); },
        },
    },
};
export var _OauthAuthorizationPage = function (args) { return (_jsx(OauthAuthorizationPage, __assign({}, args))); };
_OauthAuthorizationPage.storyName = 'OauthAuthorizationPage';
