// @rocket.chat/onboarding-ui/dist/esm/pages/CreateNewPasswordPage/CreateNewPasswordPage.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import CreateNewPasswordPage from './CreateNewPasswordPage';
export default {
    title: 'pages/CreateNewPasswordPage',
    component: CreateNewPasswordPage,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    argTypes: {
        validatePassword: { action: 'validatePassword' },
        validatePasswordConfirmation: { action: 'validatePasswordConfirmation' },
    },
    args: {
        control: { type: 'inline-radio' },
    },
};
export var _CreateNewPasswordPage = function (args) { return (_jsx(CreateNewPasswordPage, __assign({}, args))); };
_CreateNewPasswordPage.storyName = 'CreateNewPasswordPage';
