// @rocket.chat/onboarding-ui/dist/esm/forms/RequestTrialForm/RequestTrialForm.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx } from "react/jsx-runtime";
import { countries } from 'countries-list';
import RequestTrialForm from './RequestTrialForm';
export default {
    title: 'forms/RequestTrialForm',
    component: RequestTrialForm,
    parameters: {
        layout: 'centered',
        actions: { argTypesRegex: '^on.*' },
    },
    args: {
        validateEmail: function (email) {
            return email === 'rocket@rocket.chat' ? 'invalid email' : true;
        },
        organizationSizeOptions: [
            ['0', '1-10 people'],
            ['1', '11-50 people'],
            ['2', '51-100 people'],
            ['3', '101-250 people'],
            ['4', '251-500 people'],
            ['5', '501-1000 people'],
            ['6', '1001-4000 people'],
            ['7', '4001 or more people'],
        ],
        countryOptions: __spreadArray(__spreadArray([], Object.entries(countries).map(function (_a) {
            var code = _a[0], name = _a[1].name;
            return [code, name];
        }), true), [
            ['worldwide', 'Worldwide'],
        ], false),
    },
};
export var _RequestTrialForm = function (args) { return (_jsx(RequestTrialForm, __assign({}, args))); };
_RequestTrialForm.storyName = 'RequestTrialForm';
