// @rocket.chat/onboarding-ui/dist/cjs/forms/OrganizationInfoForm/OrganizationInfoForm.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._OrganizationInfoForm = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var countries_list_1 = require("countries-list");
var OrganizationInfoForm_1 = __importDefault(require("./OrganizationInfoForm"));
exports.default = {
    title: 'forms/OrganizationInfoForm',
    component: OrganizationInfoForm_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'centered',
    },
    args: {
        currentStep: 1,
        stepCount: 1,
        organizationIndustryOptions: [
            ['aerospaceDefense', 'Aerospace and Defense'],
            ['blockchain', 'Blockchain'],
            ['consulting', 'Consulting'],
            ['consumerGoods', 'Consumer Packaged Goods'],
            ['contactCenter', 'Contact Center'],
            ['education', 'Education'],
            ['entertainment', 'Entertainment'],
            ['financialServices', 'Financial Services'],
            ['gaming', 'Gaming'],
            ['healthcare', 'Healthcare'],
            ['hospitalityBusinness', 'Hospitality Businness'],
            ['insurance', 'Insurance'],
            ['itSecurity', 'IT Security'],
            ['logistics', 'Logistics'],
            ['manufacturing', 'Manufacturing'],
            ['media', 'Media'],
            ['pharmaceutical', 'Pharmaceutical'],
            ['realEstate', 'Real Estate'],
            ['religious', 'Religious'],
            ['retail', 'Retail'],
            ['socialNetwork', 'Social Network'],
            ['technologyProvider', 'Technology Provider'],
            ['technologyServices', 'Technology Services'],
            ['telecom', 'Telecom'],
            ['utilities', 'Utilities'],
            ['other', 'Other'],
        ],
        organizationSizeOptions: [
            ['0', '1-250 people'],
            ['1', '251-500 people'],
            ['2', '501-1000 people'],
            ['3', '1001-4000 people'],
            ['4', '4001 or more people'],
        ],
        countryOptions: __spreadArray(__spreadArray([], Object.entries(countries_list_1.countries).map(function (_a) {
            var code = _a[0], name = _a[1].name;
            return [code, name];
        }), true), [
            ['worldwide', 'Worldwide'],
        ], false),
    },
};
var _OrganizationInfoForm = function (args) { return ((0, jsx_runtime_1.jsx)(OrganizationInfoForm_1.default, __assign({}, args))); };
exports._OrganizationInfoForm = _OrganizationInfoForm;
exports._OrganizationInfoForm.storyName = 'OrganizationInfoForm';
