// @rocket.chat/pdf-worker/dist/templates/ChatTranscript/components/Divider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Divider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const renderer_1 = require("@react-pdf/renderer");
const colors_json_1 = __importDefault(require("@rocket.chat/fuselage-tokens/colors.json"));
const styles = renderer_1.StyleSheet.create({
    wrapper: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingVertical: 12,
        fontSize: 12,
        fontWeight: 700,
    },
    line: {
        flex: 1,
        height: 2,
        backgroundColor: colors_json_1.default.n200,
    },
    text: {
        paddingHorizontal: 8,
    },
});
const Divider = ({ divider }) => ((0, jsx_runtime_1.jsxs)(renderer_1.View, { style: styles.wrapper, wrap: false, children: [(0, jsx_runtime_1.jsx)(renderer_1.View, { style: styles.line }), (0, jsx_runtime_1.jsx)(renderer_1.Text, { style: styles.text, children: divider }), (0, jsx_runtime_1.jsx)(renderer_1.View, { style: styles.line })] }));
exports.Divider = Divider;
