// @rocket.chat/pdf-worker/dist/templates/ChatTranscript/components/MessageList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageList = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const renderer_1 = require("@react-pdf/renderer");
const typography_json_1 = require("@rocket.chat/fuselage-tokens/typography.json");
const markup_1 = require("../markup");
const Divider_1 = require("./Divider");
const Files_1 = require("./Files");
const MessageHeader_1 = require("./MessageHeader");
const Quotes_1 = require("./Quotes");
const styles = renderer_1.StyleSheet.create({
    wrapper: {
        marginBottom: 16,
        paddingHorizontal: 32,
    },
    message: {
        marginTop: 1,
        fontSize: typography_json_1.fontScales.p2.fontSize,
    },
});
const MessageList = ({ messages, invalidFileMessage }) => ((0, jsx_runtime_1.jsx)(renderer_1.View, { children: messages.map((message, index) => ((0, jsx_runtime_1.jsxs)(renderer_1.View, { style: styles.wrapper, wrap: false, children: [message.divider && (0, jsx_runtime_1.jsx)(Divider_1.Divider, { divider: message.divider }), (0, jsx_runtime_1.jsx)(MessageHeader_1.MessageHeader, { name: message.u.name || message.u.username, time: message.ts }), (0, jsx_runtime_1.jsx)(renderer_1.View, { style: styles.message, children: message.md ? (0, jsx_runtime_1.jsx)(markup_1.Markup, { tokens: message.md }) : (0, jsx_runtime_1.jsx)(renderer_1.Text, { children: message.msg }) }), message.quotes && (0, jsx_runtime_1.jsx)(Quotes_1.Quotes, { quotes: message.quotes }), message.files && (0, jsx_runtime_1.jsx)(Files_1.Files, { files: message.files, invalidMessage: invalidFileMessage })] }, index))) }));
exports.MessageList = MessageList;
