// @rocket.chat/onboarding-ui/dist/esm/pages/AdminInfoPage/AdminInfoPage.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import AdminInfoPage from './AdminInfoPage';
export default {
    title: 'pages/AdminInfoPage',
    component: AdminInfoPage,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    argTypes: {
        validateUsername: { action: 'validateUsername' },
        validateEmail: { action: 'validateEmail' },
        validatePassword: { action: 'validatePassword' },
    },
    args: {
        currentStep: 1,
        stepCount: 1,
        passwordRulesHint: 'Password rules description goes here',
        control: { type: 'inline-radio' },
    },
};
export var _AdminInfoPage = function (args) { return (_jsx(AdminInfoPage, __assign({}, args))); };
export var _CloudAdminInfoPage = function (args) { return (_jsx(AdminInfoPage, __assign({}, args))); };
_AdminInfoPage.storyName = 'AdminInfoPage';
_CloudAdminInfoPage.storyName = 'CloudAdminInfoPage';
_CloudAdminInfoPage.args = {
    title: 'Your Workspace is Ready!',
    description: 'Time to create your first admin user',
    keepPosted: true,
};
