// @rocket.chat/onboarding-ui/dist/esm/forms/RegisterServerForm/RegisterServerForm.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import RegisterServerForm from './RegisterServerForm';
export default {
    title: 'forms/RegisterServerForm',
    component: RegisterServerForm,
    parameters: {
        layout: 'centered',
        actions: { argTypesRegex: '^on.*' },
    },
    args: {
        currentStep: 1,
        stepCount: 1,
    },
};
export var _RegisterServerForm = function (args) { return (_jsx(RegisterServerForm, __assign({}, args))); };
_RegisterServerForm.storyName = 'RegisterServerForm';
export var _RegisterServerFormOffline = function (args) { return (_jsx(RegisterServerForm, __assign({}, args, { offline: true }))); };
_RegisterServerFormOffline.storyName = 'RegisterServerFormOffline';
