// @rocket.chat/onboarding-ui/dist/esm/pages/RegisterServerPage/RegisterServerPage.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import RegisterServerPage from './RegisterServerPage';
export default {
    title: 'pages/RegisterServerPage',
    component: RegisterServerPage,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    args: {
        currentStep: 1,
        stepCount: 1,
        termsHref: '#',
        policyHref: '#',
    },
};
export var _RegisterServerPage = function (args) { return (_jsx(RegisterServerPage, __assign({}, args))); };
_RegisterServerPage.storyName = 'Online';
export var _RegisterServerPageOffline = function (args) { return (_jsx(RegisterServerPage, __assign({}, args, { offline: true }))); };
_RegisterServerPageOffline.storyName = 'Offline';
