// @rocket.chat/web-ui-registration/dist/LoginForm.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { FieldGroup, TextInput, Field, FieldLabel, FieldRow, FieldError, FieldLink, PasswordInput, ButtonGroup, Button, Callout, } from '@rocket.chat/fuselage';
import { useUniqueId } from '@rocket.chat/fuselage-hooks';
import { Form, ActionLink } from '@rocket.chat/layout';
import { useLoginWithPassword, useSetting } from '@rocket.chat/ui-contexts';
import { useMutation } from '@tanstack/react-query';
import { useEffect, useRef, useState } from 'react';
import { useForm } from 'react-hook-form';
import { Trans, useTranslation } from 'react-i18next';
import EmailConfirmationForm from './EmailConfirmationForm';
import LoginServices from './LoginServices';
const LOGIN_SUBMIT_ERRORS = {
    'error-user-is-not-activated': {
        type: 'warning',
        i18n: 'registration.page.registration.waitActivationWarning',
    },
    'error-app-user-is-not-allowed-to-login': {
        type: 'danger',
        i18n: 'registration.page.login.errors.AppUserNotAllowedToLogin',
    },
    'user-not-found': {
        type: 'danger',
        i18n: 'registration.page.login.errors.wrongCredentials',
    },
    'error-login-blocked-for-ip': {
        type: 'danger',
        i18n: 'registration.page.login.errors.loginBlockedForIp',
    },
    'error-login-blocked-for-user': {
        type: 'danger',
        i18n: 'registration.page.login.errors.loginBlockedForUser',
    },
    'error-license-user-limit-reached': {
        type: 'warning',
        i18n: 'registration.page.login.errors.licenseUserLimitReached',
    },
    'error-invalid-email': {
        type: 'danger',
        i18n: 'registration.page.login.errors.invalidEmail',
    },
};
export const LoginForm = ({ setLoginRoute }) => {
    var _a, _b, _c;
    const { register, handleSubmit, setError, clearErrors, getValues, formState: { errors }, } = useForm({
        mode: 'onBlur',
    });
    const { t } = useTranslation();
    const formLabelId = useUniqueId();
    const [errorOnSubmit, setErrorOnSubmit] = useState(undefined);
    const isResetPasswordAllowed = useSetting('Accounts_PasswordReset');
    const login = useLoginWithPassword();
    const showFormLogin = useSetting('Accounts_ShowFormLogin');
    const usernameOrEmailPlaceholder = String(useSetting('Accounts_EmailOrUsernamePlaceholder'));
    const passwordPlaceholder = String(useSetting('Accounts_PasswordPlaceholder'));
    const loginMutation = useMutation({
        mutationFn: (formData) => {
            return login(formData.username, formData.password);
        },
        onError: (error) => {
            if ([error.error, error.errorType].includes('error-invalid-email')) {
                setError('username', { type: 'invalid-email', message: t('registration.page.login.errors.invalidEmail') });
            }
            if ('error' in error && error.error !== 403) {
                setErrorOnSubmit(error.error);
                return;
            }
            setErrorOnSubmit('user-not-found');
        },
    });
    const usernameId = useUniqueId();
    const passwordId = useUniqueId();
    const loginFormRef = useRef(null);
    useEffect(() => {
        if (loginFormRef.current) {
            loginFormRef.current.focus();
        }
    }, [errorOnSubmit]);
    const renderErrorOnSubmit = (error) => {
        const { type, i18n } = LOGIN_SUBMIT_ERRORS[error];
        return (_jsx(Callout, { id: `${usernameId}-error`, "aria-live": 'assertive', type: type, children: t(i18n) }));
    };
    if (((_a = errors.username) === null || _a === void 0 ? void 0 : _a.type) === 'invalid-email') {
        return _jsx(EmailConfirmationForm, { onBackToLogin: () => clearErrors('username'), email: getValues('username') });
    }
    return (_jsxs(Form, { tabIndex: -1, ref: loginFormRef, "aria-labelledby": formLabelId, "aria-describedby": 'welcomeTitle', onSubmit: handleSubmit((data) => __awaiter(void 0, void 0, void 0, function* () { return loginMutation.mutate(data); })), children: [_jsx(Form.Header, { children: _jsx(Form.Title, { id: formLabelId, children: t('registration.component.login') }) }), showFormLogin && (_jsxs(_Fragment, { children: [_jsxs(Form.Container, { children: [_jsxs(FieldGroup, { disabled: loginMutation.isLoading, children: [_jsxs(Field, { children: [_jsx(FieldLabel, { required: true, htmlFor: usernameId, children: t('registration.component.form.emailOrUsername') }), _jsx(FieldRow, { children: _jsx(TextInput, Object.assign({}, register('username', {
                                                    required: t('registration.component.form.requiredField'),
                                                }), { placeholder: usernameOrEmailPlaceholder || t('registration.component.form.emailPlaceholder'), error: (_b = errors.username) === null || _b === void 0 ? void 0 : _b.message, "aria-invalid": errors.username || errorOnSubmit ? 'true' : 'false', "aria-describedby": `${usernameId}-error`, id: usernameId })) }), errors.username && (_jsx(FieldError, { "aria-live": 'assertive', id: `${usernameId}-error`, children: errors.username.message }))] }), _jsxs(Field, { children: [_jsx(FieldLabel, { required: true, htmlFor: passwordId, children: t('registration.component.form.password') }), _jsx(FieldRow, { children: _jsx(PasswordInput, Object.assign({}, register('password', {
                                                    required: t('registration.component.form.requiredField'),
                                                }), { placeholder: passwordPlaceholder, error: (_c = errors.password) === null || _c === void 0 ? void 0 : _c.message, "aria-invalid": errors.password || errorOnSubmit ? 'true' : 'false', "aria-describedby": `${passwordId}-error`, id: passwordId })) }), errors.password && (_jsx(FieldError, { "aria-live": 'assertive', id: `${passwordId}-error`, children: errors.password.message })), isResetPasswordAllowed && (_jsx(FieldRow, { justifyContent: 'end', children: _jsx(FieldLink, { href: '#', onClick: (e) => {
                                                        e.preventDefault();
                                                        setLoginRoute('reset-password');
                                                    }, children: _jsx(Trans, { i18nKey: 'registration.page.login.forgot', children: "Forgot your password?" }) }) }))] })] }), errorOnSubmit && _jsx(FieldGroup, { disabled: loginMutation.isLoading, children: renderErrorOnSubmit(errorOnSubmit) })] }), _jsxs(Form.Footer, { children: [_jsx(ButtonGroup, { stretch: true, children: _jsx(Button, { loading: loginMutation.isLoading, type: 'submit', primary: true, children: t('registration.component.login') }) }), _jsx("p", { children: _jsxs(Trans, { i18nKey: 'registration.page.login.register', children: ["New here? ", _jsx(ActionLink, { onClick: () => setLoginRoute('register'), children: "Create an account" })] }) })] })] })), _jsx(LoginServices, { disabled: loginMutation.isLoading, setError: setErrorOnSubmit })] }));
};
export default LoginForm;
