// @rocket.chat/onboarding-ui/dist/esm/forms/RegisterOfflineForm/RegisterOfflineForm.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import RegisterOfflineForm from './RegisterOfflineForm';
export default {
    title: 'forms/RegisterOfflineForm',
    component: RegisterOfflineForm,
    parameters: {
        layout: 'centered',
        actions: { argTypesRegex: '^on.*' },
    },
    args: {
        clientKey: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
    },
};
export var _RegisterOfflineForm = function (args) { return (_jsx(RegisterOfflineForm, __assign({}, args))); };
