// @rocket.chat/rest-typings/dist/v1/banners.js
import Ajv from 'ajv';
const ajv = new Ajv({
    coerceTypes: true,
});
const BannersGetNewSchema = {
    type: 'object',
    properties: {
        platform: {
            type: 'string',
            enum: ['1', '2'],
        },
        bid: {
            type: 'string',
        },
    },
    required: ['platform', 'bid'],
    additionalProperties: false,
};
export const isBannersGetNewProps = ajv.compile(BannersGetNewSchema);
const BannersIdSchema = {
    type: 'object',
    properties: {
        platform: {
            type: 'string',
        },
    },
    required: ['platform'],
    additionalProperties: false,
};
export const isBannersIdProps = ajv.compile(BannersIdSchema);
const BannersSchema = {
    type: 'object',
    properties: {
        platform: {
            type: 'string',
        },
    },
    required: ['platform'],
    additionalProperties: false,
};
export const isBannersProps = ajv.compile(BannersSchema);
const BannersDismissSchema = {
    type: 'object',
    properties: {
        bannerId: {
            type: 'string',
        },
    },
    required: ['bannerId'],
    additionalProperties: false,
};
export const isBannersDismissProps = ajv.compile(BannersDismissSchema);
