// @rocket.chat/rest-typings/dist/v1/e2e.js
import Ajv from 'ajv';
const ajv = new Ajv({
    coerceTypes: true,
});
const E2eSetUserPublicAndPrivateKeysSchema = {
    type: 'object',
    properties: {
        public_key: {
            type: 'string',
        },
        private_key: {
            type: 'string',
        },
    },
    required: ['public_key', 'private_key'],
    additionalProperties: false,
};
export const isE2eSetUserPublicAndPrivateKeysProps = ajv.compile(E2eSetUserPublicAndPrivateKeysSchema);
const E2eGetUsersOfRoomWithoutKeySchema = {
    type: 'object',
    properties: {
        rid: {
            type: 'string',
        },
    },
    required: ['rid'],
    additionalProperties: false,
};
export const isE2eGetUsersOfRoomWithoutKeyProps = ajv.compile(E2eGetUsersOfRoomWithoutKeySchema);
const E2eUpdateGroupKeySchema = {
    type: 'object',
    properties: {
        uid: {
            type: 'string',
        },
        rid: {
            type: 'string',
        },
        key: {
            type: 'string',
        },
    },
    required: ['uid', 'rid', 'key'],
    additionalProperties: false,
};
export const isE2eUpdateGroupKeyProps = ajv.compile(E2eUpdateGroupKeySchema);
const E2eSetRoomKeyIdSchema = {
    type: 'object',
    properties: {
        rid: {
            type: 'string',
        },
        keyID: {
            type: 'string',
        },
    },
    required: ['rid', 'keyID'],
    additionalProperties: false,
};
export const isE2eSetRoomKeyIdProps = ajv.compile(E2eSetRoomKeyIdSchema);
