// @rocket.chat/rest-typings/dist/v1/permissions.js
import Ajv from 'ajv';
const ajv = new Ajv({
    coerceTypes: true,
});
const permissionListAllSchema = {
    type: 'object',
    properties: {
        updatedSince: {
            type: 'string',
            nullable: true,
        },
    },
    required: [],
    additionalProperties: false,
};
export const isPermissionsListAll = ajv.compile(permissionListAllSchema);
const permissionUpdatePropsSchema = {
    type: 'object',
    properties: {
        permissions: {
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    _id: { type: 'string' },
                    roles: {
                        type: 'array',
                        items: { type: 'string' },
                        uniqueItems: true,
                    },
                },
                additionalProperties: false,
                required: ['_id', 'roles'],
            },
        },
    },
    required: ['permissions'],
    additionalProperties: false,
};
export const isBodyParamsValidPermissionUpdate = ajv.compile(permissionUpdatePropsSchema);
