// @rocket.chat/onboarding-ui/dist/esm/pages/CreateNewAccountPage/CreateNewAccountPage.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import CreateNewAccountPage from './CreateNewAccountPage';
export default {
    title: 'pages/CreateNewAccountPage',
    component: CreateNewAccountPage,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'fullscreen',
    },
    argTypes: {
        validateEmail: { action: 'validateEmail' },
        validatePassword: { action: 'validatePassword' },
        validateConfirmationPassword: { action: 'validateConfirmationPassword' },
    },
    args: {
        control: { type: 'inline-radio' },
    },
};
export var _CreateNewAccountPage = function (args) { return (_jsx(CreateNewAccountPage, __assign({}, args))); };
