// @rocket.chat/onboarding-ui/dist/cjs/forms/AdminInfoForm/AdminInfoForm.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._AdminInfoForm = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var AdminInfoForm_1 = __importDefault(require("./AdminInfoForm"));
exports.default = {
    title: 'forms/AdminInfoForm',
    component: AdminInfoForm_1.default,
    parameters: {
        actions: { argTypesRegex: '^on.*' },
        layout: 'centered',
    },
    argTypes: {
        validateUsername: { action: 'validateUsername' },
        validateEmail: { action: 'validateEmail' },
        validatePassword: { action: 'validatePassword' },
    },
    args: {
        currentStep: 1,
        stepCount: 1,
        passwordRulesHint: 'Password rules description goes here',
    },
};
var _AdminInfoForm = function (args) { return ((0, jsx_runtime_1.jsx)(AdminInfoForm_1.default, __assign({}, args))); };
exports._AdminInfoForm = _AdminInfoForm;
exports._AdminInfoForm.storyName = 'AdminInfoForm';
